pool/app/custom_field/custom_field.ml

```diff
@@ -29,101 +33,169 @@ let find_options_by_field pool id =
 let find_group = Repo_group.find
 let find_groups_by_model = Repo_group.find_by_model
 
-let validate_htmx value (m : Public.t) =
+module Repo = struct
+  module Id = struct
+    include Pool_common.Repo.Id
+  end
+
+  module SelectOption = struct
+    module Id = struct
+      include Repo_entity.Option.Id
+    end
+  end
+end
+
+let create_answer is_admin answer new_value =
+  let id = Answer.id_opt answer in
+  let value = CCOption.bind answer (fun a -> a.Answer.value) in
+  (match is_admin with
+   | true -> Answer.create ?id ~admin_value:new_value value
+   | false -> Answer.create ?id (Some new_value))
+  |> CCOption.pure
+;;
+
+let validate_htmx ~is_admin value (m : Public.t) =
   let open Public in
   let open CCResult.Infix in
-  let open CCFun in
   let no_value = Error Pool_common.Message.NoValue in
-  let required = Public.required m in
+  (* Allow admins to reset required answers *)
+  let required = Public.required m && not is_admin in
   let single_value =
     value
     |> CCList.head_opt
-    |> flip CCOption.bind (fun v ->
+    |> CCFun.flip CCOption.bind (fun v ->
          if CCString.is_empty v then None else Some v)
   in
   let go validation value = validation |> fst |> fun rule -> rule value in
   match m with
   | Boolean (public, answer) ->
     (* TODO: Find UI way to do this *)
     let to_field a = Public.Boolean (public, a) |> CCResult.return in
-    let id = Answer.id_opt answer in
     (match single_value, required with
      | Some value, _ ->
        value
        |> Utils.Bool.of_string
-       |> Answer.create ?id
-       |> CCOption.pure
+       |> create_answer is_admin answer
        |> to_field
      | None, false -> to_field None
      | None, true -> no_value)
-  | MultiSelect (public, options, _) ->
+  | MultiSelect (public, options, answer) ->
     let to_field a = Public.MultiSelect (public, options, a) in
     (match value, required with
      | [], true -> no_value
      | vals, _ ->
        let open SelectOption in
-       vals
-       |> CCList.map (fun value ->
-            let id = value |> Id.of_string in
-            CCList.find_opt (fun option -> Id.equal option.Public.id id) options
-            |> CCOption.to_result
-                 Pool_common.Message.(Invalid Field.CustomFieldOption)
-            >|= Answer.create)
-       |> CCList.all_ok
-       >|= to_field)
+       let a =
+         vals
+         |> CCList.map (fun value ->
+              CCList.find_opt
+                (fun option ->
+                  Id.equal option.Public.id (value |> Id.of_string))
```

Comment:
Shorten find function

```suggestion
                (fun { Public.id; _ } -> Id.equal id (value |> Id.of_string))
```

I thought about using "CCList.mem" or "CCList.find_map", but I guess this is easiest.

Reply:
👍